src/coreclr/jit/importer.cpp

```diff
@@ -6929,30 +6929,33 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                     }
 
                     // If we see a local being assigned the result of a GDV-inlineable
-                    // IEnumerable<T>.GetEnumerator, keep track of both the local and the call.
+                    // GetEnumerator call, keep track of both the local and the call.
                     //
                     if (op1->OperIs(GT_RET_EXPR))
                     {
-                        JITDUMP(".... checking for GDV of IEnumerable<T>...\n");
+                        JITDUMP(".... checking for GDV returning IEnumerator<T>...\n");
 
-                        GenTreeCall* const call = op1->AsRetExpr()->gtInlineCandidate;
-                        NamedIntrinsic     ni   = NI_Illegal;
+                        bool                 isEnumeratorT = false;
+                        GenTreeCall* const   call          = op1->AsRetExpr()->gtInlineCandidate;
+                        bool                 isExact       = false;
+                        bool                 isNonNull     = false;
+                        CORINFO_CLASS_HANDLE retCls        = gtGetClassHandle(call, &isExact, &isNonNull);
 
-                        // TODO -- handle CT_INDIRECT virtuals here too
-                        // but we don't have the right method handle
-                        //
-                        if (call->gtCallType == CT_USER_FUNC)
-                        {
-                            ni = lookupNamedIntrinsic(call->gtCallMethHnd);
-                        }
-                        else if (call->IsGuardedDevirtualizationCandidate())
+                        if ((retCls != NO_CLASS_HANDLE) && info.compCompHnd->isIntrinsicType(retCls))
                         {
-                            JITDUMP("No GDV IEnumerable<T> check for [%06u]\n", dspTreeID(call));
+                            const char* namespaceName;
+                            const char* className = info.compCompHnd->getClassNameFromMetadata(retCls, &namespaceName);
+
+                            if ((strcmp(namespaceName, "System.Collections.Generic") == 0) &&
+                                (strcmp(className, "IEnumerator`1") == 0))
```

Comment:
Perhaps move this check in `lookupNamedIntrinsic` and rename `NI_System_Collections_Generic_IEnumerable_GetEnumerator` since the constant is now unsued.